# Before your start:
- Read the README.md file
- Comment as much as you can and use the resources in the README.md file
- Happy learning!

In [1]:
#Import your libraries
import pandas as pd
import numpy as np

import matplotlib.pyplot as plt
import seaborn as sns

from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.metrics import r2_score, mean_squared_error, mean_absolute_error


# Challenge 1 - Import and Describe the Dataset

#### In this challenge we will use the `austin_weather` data. 

#### First, import it into a data frame called `austin`. 

In [2]:
# Your code here
ruta_archivo_csv = 'austin_weather.csv'
data = pd.read_csv(ruta_archivo_csv)

#### Next, describe the dataset you have loaded: 
- Look at the variables and their types
- Examine the descriptive statistics of the numeric variables 
- Look at the first five rows of all variables to evaluate the categorical variables as well

In [3]:
data.shape

(1319, 21)

In [4]:
data.info()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1319 entries, 0 to 1318
Data columns (total 21 columns):
 #   Column                      Non-Null Count  Dtype 
---  ------                      --------------  ----- 
 0   Date                        1319 non-null   object
 1   TempHighF                   1319 non-null   int64 
 2   TempAvgF                    1319 non-null   int64 
 3   TempLowF                    1319 non-null   int64 
 4   DewPointHighF               1319 non-null   object
 5   DewPointAvgF                1319 non-null   object
 6   DewPointLowF                1319 non-null   object
 7   HumidityHighPercent         1319 non-null   object
 8   HumidityAvgPercent          1319 non-null   object
 9   HumidityLowPercent          1319 non-null   object
 10  SeaLevelPressureHighInches  1319 non-null   object
 11  SeaLevelPressureAvgInches   1319 non-null   object
 12  SeaLevelPressureLowInches   1319 non-null   object
 13  VisibilityHighMiles         1319 non-null   obje

In [5]:
data.columns


Index(['Date', 'TempHighF', 'TempAvgF', 'TempLowF', 'DewPointHighF',
       'DewPointAvgF', 'DewPointLowF', 'HumidityHighPercent',
       'HumidityAvgPercent', 'HumidityLowPercent',
       'SeaLevelPressureHighInches', 'SeaLevelPressureAvgInches',
       'SeaLevelPressureLowInches', 'VisibilityHighMiles',
       'VisibilityAvgMiles', 'VisibilityLowMiles', 'WindHighMPH', 'WindAvgMPH',
       'WindGustMPH', 'PrecipitationSumInches', 'Events'],
      dtype='object')

In [6]:
data.sample(3)

Date  TempHighF  TempAvgF  TempLowF DewPointHighF DewPointAvgF  \
376   2015-01-01         40        37        34            39           35   
59    2014-02-18         82        69        55            63           53   
1085  2016-12-10         52        44        35            50           37   

     DewPointLowF HumidityHighPercent HumidityAvgPercent HumidityLowPercent  \
376            29                 100                 82                 64   
59             32                  78                 55                 32   
1085           27                  96                 75                 54   

      ... SeaLevelPressureAvgInches SeaLevelPressureLowInches  \
376   ...                     30.25                     30.15   
59    ...                        30                     29.89   
1085  ...                     30.37                      30.2   

     VisibilityHighMiles VisibilityAvgMiles VisibilityLowMiles WindHighMPH  \
376                   10                  3                  1           9   
59                    10                 10                  9          13   
1085                  10                  8                  2           9   

     WindAvgMPH WindGustMPH PrecipitationSumInches               Events  
376           5          13                   0.63  Rain , Thunderstorm  
59            4          20                      0                       
1085          3          15                      0                       

[3 rows x 21 columns]

In [7]:
df = data.copy()

In [8]:
df.dtypes

Date                          object
TempHighF                      int64
TempAvgF                       int64
TempLowF                       int64
DewPointHighF                 object
DewPointAvgF                  object
DewPointLowF                  object
HumidityHighPercent           object
HumidityAvgPercent            object
HumidityLowPercent            object
SeaLevelPressureHighInches    object
SeaLevelPressureAvgInches     object
SeaLevelPressureLowInches     object
VisibilityHighMiles           object
VisibilityAvgMiles            object
VisibilityLowMiles            object
WindHighMPH                   object
WindAvgMPH                    object
WindGustMPH                   object
PrecipitationSumInches        object
Events                        object
dtype: object

In [9]:
df.isna().sum()

Date                          0
TempHighF                     0
TempAvgF                      0
TempLowF                      0
DewPointHighF                 0
DewPointAvgF                  0
DewPointLowF                  0
HumidityHighPercent           0
HumidityAvgPercent            0
HumidityLowPercent            0
SeaLevelPressureHighInches    0
SeaLevelPressureAvgInches     0
SeaLevelPressureLowInches     0
VisibilityHighMiles           0
VisibilityAvgMiles            0
VisibilityLowMiles            0
WindHighMPH                   0
WindAvgMPH                    0
WindGustMPH                   0
PrecipitationSumInches        0
Events                        0
dtype: int64

In [10]:
df.duplicated().sum()

0

In [11]:
df.eq(" ").sum()

Date                            0
TempHighF                       0
TempAvgF                        0
TempLowF                        0
DewPointHighF                   0
DewPointAvgF                    0
DewPointLowF                    0
HumidityHighPercent             0
HumidityAvgPercent              0
HumidityLowPercent              0
SeaLevelPressureHighInches      0
SeaLevelPressureAvgInches       0
SeaLevelPressureLowInches       0
VisibilityHighMiles             0
VisibilityAvgMiles              0
VisibilityLowMiles              0
WindHighMPH                     0
WindAvgMPH                      0
WindGustMPH                     0
PrecipitationSumInches          0
Events                        903
dtype: int64

In [12]:
df.describe().T

count       mean        std   min   25%   50%   75%    max
TempHighF  1319.0  80.862775  14.766523  32.0  72.0  83.0  92.0  107.0
TempAvgF   1319.0  70.642911  14.045904  29.0  62.0  73.0  83.0   93.0
TempLowF   1319.0  59.902957  14.190648  19.0  49.0  63.0  73.0   81.0

#### Given the information you have learned from examining the dataset, write down three insights about the data in a markdown cell below

#### Your Insights:

1. There are 21 variables in the dataset. 3 of them are numeric and the rest contain some text.

2. The average temperature in Austin ranged between around 70 degrees F and around 93 degrees F. The highest temperature observed during this period was 107 degrees F and the lowest was 19 degrees F.

3. When we look at the head function, we see that a lot of variables contain numeric data even though these columns are of object type. This means we might have to do some data cleansing.


#### Let's examine the DewPointAvgF variable by using the `unique()` function to list all unique values in this dataframe.

Describe what you find in a markdown cell below the code. What did you notice? What do you think made Pandas to treat this column as *object* instead of *int64*? 

In [13]:
valores_unicos = df['DewPointAvgF'].unique()
valores_unicos

array(['49', '36', '27', '28', '40', '39', '41', '26', '42', '22', '48',
       '32', '8', '11', '45', '55', '61', '37', '47', '25', '23', '20',
       '33', '30', '29', '17', '14', '13', '54', '59', '15', '24', '34',
       '35', '57', '50', '53', '60', '46', '56', '51', '31', '38', '62',
       '43', '63', '64', '67', '66', '58', '70', '68', '65', '69', '71',
       '72', '-', '73', '74', '21', '44', '52', '12', '75', '76', '18'],
      dtype=object)

The following is a list of columns misrepresented as `object`. Use this list to convert the columns to numeric using the `pandas.to_numeric` function in the next cell. If you encounter errors in converting strings to numeric values, you need to catch those errors and force the conversion by supplying `errors='coerce'` as an argument for `pandas.to_numeric`. Coercing will replace non-convertable elements with `NaN` which represents an undefined numeric value. This makes it possible for us to conveniently handle missing values in subsequent data processing.

*Hint: you may use a loop to change one column at a time but it is more efficient to use `apply`.*

In [14]:
wrong_type_columns = ['DewPointHighF', 'DewPointAvgF', 'DewPointLowF', 'HumidityHighPercent', 
                      'HumidityAvgPercent', 'HumidityLowPercent', 'SeaLevelPressureHighInches', 
                      'SeaLevelPressureAvgInches' ,'SeaLevelPressureLowInches', 'VisibilityHighMiles',
                      'VisibilityAvgMiles', 'VisibilityLowMiles', 'WindHighMPH', 'WindAvgMPH', 
                      'WindGustMPH', 'PrecipitationSumInches']

In [15]:


# Aplicar la conversión a todas las columnas
df[wrong_type_columns] = df[wrong_type_columns].apply(pd.to_numeric, errors='coerce')


In [16]:
df.dtypes

Date                           object
TempHighF                       int64
TempAvgF                        int64
TempLowF                        int64
DewPointHighF                 float64
DewPointAvgF                  float64
DewPointLowF                  float64
HumidityHighPercent           float64
HumidityAvgPercent            float64
HumidityLowPercent            float64
SeaLevelPressureHighInches    float64
SeaLevelPressureAvgInches     float64
SeaLevelPressureLowInches     float64
VisibilityHighMiles           float64
VisibilityAvgMiles            float64
VisibilityLowMiles            float64
WindHighMPH                   float64
WindAvgMPH                    float64
WindGustMPH                   float64
PrecipitationSumInches        float64
Events                         object
dtype: object

#### Check if your code has worked by printing the data types again. You should see only two `object` columns (`Date` and `Events`) now. All other columns should be `int64` or `float64`.

In [17]:
# Your code here


# Challenge 2 - Handle the Missing Data

#### Now that we have fixed the type mismatch, let's address the missing data.

By coercing the columns to numeric, we have created `NaN` for each cell containing characters. We should choose a strategy to address these missing data.

The first step is to examine how many rows contain missing data.

We check how much missing data we have by applying the `.isnull()` function to our dataset. To find the rows with missing data in any of its cells, we apply `.any(axis=1)` to the function. `austin.isnull().any(axis=1)` will return a column containing true if the row contains at least one missing value and false otherwise. Therefore we must subset our dataframe with this column. This will give us all rows with at least one missing value. 

#### In the next cell, identify all rows containing at least one missing value. Assign the dataframes with missing values to a variable called `missing_values`.

In [18]:
missing_values = df[df.isnull().any(axis=1)]

missing_values.shape

(136, 21)

In [19]:
df.shape

(1319, 21)

In [20]:
missing_values.sample(3)

Date  TempHighF  TempAvgF  TempLowF  DewPointHighF  DewPointAvgF  \
1209  2017-04-13         82        75        67           66.0          64.0   
174   2014-06-13         89        79        68            NaN           NaN   
188   2014-06-27         91        83        75           74.0          72.0   

      DewPointLowF  HumidityHighPercent  HumidityAvgPercent  \
1209          58.0                100.0                77.0   
174            NaN                 95.0                75.0   
188           70.0                 88.0                69.0   

      HumidityLowPercent  ...  SeaLevelPressureAvgInches  \
1209                54.0  ...                      30.14   
174                 55.0  ...                      29.95   
188                 49.0  ...                      29.87   

      SeaLevelPressureLowInches  VisibilityHighMiles  VisibilityAvgMiles  \
1209                      30.04                 10.0                 9.0   
174                       29.87                  NaN                 NaN   
188                       29.79                 10.0                10.0   

      VisibilityLowMiles  WindHighMPH  WindAvgMPH  WindGustMPH  \
1209                 5.0         14.0         6.0         23.0   
174                  NaN         15.0         4.0         22.0   
188                  8.0         17.0         8.0         32.0   

      PrecipitationSumInches  Events  
1209                     NaN    Rain  
174                      0.0          
188                      NaN          

[3 rows x 21 columns]

In [21]:
df.sample(3)

Date  TempHighF  TempAvgF  TempLowF  DewPointHighF  DewPointAvgF  \
796   2016-02-25         69        54        38           34.0          31.0   
1203  2017-04-07         81        67        52           51.0          48.0   
1142  2017-02-05         72        62        51           64.0          59.0   

      DewPointLowF  HumidityHighPercent  HumidityAvgPercent  \
796           27.0                 79.0                53.0   
1203          44.0                 77.0                55.0   
1142          51.0                100.0                89.0   

      HumidityLowPercent  ...  SeaLevelPressureAvgInches  \
796                 26.0  ...                      30.31   
1203                32.0  ...                      30.17   
1142                78.0  ...                      30.06   

      SeaLevelPressureLowInches  VisibilityHighMiles  VisibilityAvgMiles  \
796                       30.21                 10.0                10.0   
1203                      30.07                 10.0                10.0   
1142                      29.98                 10.0                 6.0   

      VisibilityLowMiles  WindHighMPH  WindAvgMPH  WindGustMPH  \
796                 10.0         13.0         5.0         19.0   
1203                10.0         12.0         4.0         20.0   
1142                 0.0         12.0         3.0         19.0   

      PrecipitationSumInches  Events  
796                      0.0          
1203                     0.0          
1142                     NaN     Fog  

[3 rows x 21 columns]

There are multiple strategies to handle missing data. Below lists the most common ones data scientists use:

* Removing all rows or all columns containing missing data. This is the simplest strategy. It may work in some cases but not others.

* Filling all missing values with a placeholder value. 
    * For categorical data, `0`, `-1`, and `9999` are some commonly used placeholder values. 
    * For continuous data, some may opt to fill all missing data with the mean. This strategy is not optimal since it can increase the fit of the model.

* Filling the values using some algorithm. 

#### In our case, we will use a hybrid approach which is to first remove the data that contain most missing values then fill in the rest of the missing values with the *linear interpolation* algorithm.

#### Next, count the number of rows of `austin` and `missing_values`.

In [22]:
# Your code here
austin = df.copy()

#### Calculate the ratio of missing rows to total rows

In [23]:
# Your code here
len(missing_values)

136

In [24]:
len(austin)

1319

In [25]:
ratio = len(missing_values)/len(austin)*100

In [26]:
ratio

10.310841546626232

As you can see, there is a large proportion of missing data (over 10%). Perhaps we should evaluate which columns have the most missing data and remove those columns. For the remaining columns, we will perform a linear approximation of the missing data.

We can find the number of missing rows in each column using the `.isna()` function. We then chain the `.sum` function to the `.isna()` function and find the number of missing rows per column

In [27]:
missing_rows_per_column = austin.isna().sum()
missing_rows_per_column

Date                            0
TempHighF                       0
TempAvgF                        0
TempLowF                        0
DewPointHighF                   7
DewPointAvgF                    7
DewPointLowF                    7
HumidityHighPercent             2
HumidityAvgPercent              2
HumidityLowPercent              2
SeaLevelPressureHighInches      3
SeaLevelPressureAvgInches       3
SeaLevelPressureLowInches       3
VisibilityHighMiles            12
VisibilityAvgMiles             12
VisibilityLowMiles             12
WindHighMPH                     2
WindAvgMPH                      2
WindGustMPH                     4
PrecipitationSumInches        124
Events                          0
dtype: int64

#### As you can see from the output, the majority of missing data is in one column called `PrecipitationSumInches`. What's the number of missing values in this column in ratio to its total number of rows?

In [28]:
ratio_PrecipitationSumInches= 124/len(austin)*100
ratio_PrecipitationSumInches

9.401061410159212

Almost 10% data missing! Therefore, we prefer to remove this column instead of filling its missing values. 

#### Remove this column from `austin` using the `.drop()` function. Use the `inplace=True` argument.

*Hints:*

* By supplying `inplace=True` to `drop()`, the original dataframe object will be changed in place and the function will return `None`. In contrast, if you don't supply `inplace=True`, which is equivalent to supplying `inplace=False` because `False` is the default value, the original dataframe object will be kept and the function returns a copy of the transformed dataframe object. In the latter case, you'll have to assign the returned object back to your variable.

* Also, since you are dropping a column instead of a row, you'll need to supply `axis=1` to `drop()`.

[Reference for `pandas.DataFrame.drop`](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.DataFrame.drop.html)

In [29]:
austin.drop("PrecipitationSumInches", axis=1, inplace=True)


In [30]:
austin.columns

Index(['Date', 'TempHighF', 'TempAvgF', 'TempLowF', 'DewPointHighF',
       'DewPointAvgF', 'DewPointLowF', 'HumidityHighPercent',
       'HumidityAvgPercent', 'HumidityLowPercent',
       'SeaLevelPressureHighInches', 'SeaLevelPressureAvgInches',
       'SeaLevelPressureLowInches', 'VisibilityHighMiles',
       'VisibilityAvgMiles', 'VisibilityLowMiles', 'WindHighMPH', 'WindAvgMPH',
       'WindGustMPH', 'Events'],
      dtype='object')

#### Next we will perform linear interpolation of the missing data.

This means that we will use a linear algorithm to estimate the missing data. Linear interpolation assumes that there is a straight line between the points and the missing point will fall on that line. This is a good enough approximation for weather related data. Weather related data is typically a time series. Therefore, we do not want to drop rows from our data if possible. It is prefereable to estimate the missing values rather than remove the rows. However, if you have data from a single point in time, perhaps a better solution would be to remove the rows. 

If you would like to read more about linear interpolation, you can do so [here](https://en.wikipedia.org/wiki/Linear_interpolation).

In the following cell, use the `.interpolate()` function on the entire dataframe. This time pass the `inplace=False` argument to the function and assign the interpolated dataframe to a new variable called `austin_fixed` so that we can compare with `austin`.

In [31]:
austin_fixed = austin.interpolate(inplace=False)

#### Check to make sure `austin_fixed` contains no missing data. Also check `austin` - it still contains missing data.

In [32]:
austin_fixed.sample(3)

Date  TempHighF  TempAvgF  TempLowF  DewPointHighF  DewPointAvgF  \
35    2014-01-25         70        48        26           38.0          27.0   
730   2015-12-21         76        65        54           64.0          61.0   
1257  2017-05-31         85        77        68           70.0          68.0   

      DewPointLowF  HumidityHighPercent  HumidityAvgPercent  \
35            22.0                 85.0                53.0   
730           51.0                 93.0                79.0   
1257          62.0                 97.0                74.0   

      HumidityLowPercent  SeaLevelPressureHighInches  \
35                  20.0                       30.38   
730                 64.0                       30.01   
1257                51.0                       30.08   

      SeaLevelPressureAvgInches  SeaLevelPressureLowInches  \
35                        30.19                      30.02   
730                       29.96                      29.88   
1257                      30.01                      29.94   

      VisibilityHighMiles  VisibilityAvgMiles  VisibilityLowMiles  \
35                   10.0                10.0                 8.0   
730                  10.0                 9.0                 5.0   
1257                 10.0                10.0                 9.0   

      WindHighMPH  WindAvgMPH  WindGustMPH Events  
35            8.0         1.0         14.0         
730           8.0         2.0         15.0         
1257         17.0         2.0         26.0   Rain

In [33]:
austin_fixed.isna().sum()

Date                          0
TempHighF                     0
TempAvgF                      0
TempLowF                      0
DewPointHighF                 0
DewPointAvgF                  0
DewPointLowF                  0
HumidityHighPercent           0
HumidityAvgPercent            0
HumidityLowPercent            0
SeaLevelPressureHighInches    0
SeaLevelPressureAvgInches     0
SeaLevelPressureLowInches     0
VisibilityHighMiles           0
VisibilityAvgMiles            0
VisibilityLowMiles            0
WindHighMPH                   0
WindAvgMPH                    0
WindGustMPH                   0
Events                        0
dtype: int64

In [34]:
austin.isna().sum()

Date                           0
TempHighF                      0
TempAvgF                       0
TempLowF                       0
DewPointHighF                  7
DewPointAvgF                   7
DewPointLowF                   7
HumidityHighPercent            2
HumidityAvgPercent             2
HumidityLowPercent             2
SeaLevelPressureHighInches     3
SeaLevelPressureAvgInches      3
SeaLevelPressureLowInches      3
VisibilityHighMiles           12
VisibilityAvgMiles            12
VisibilityLowMiles            12
WindHighMPH                    2
WindAvgMPH                     2
WindGustMPH                    4
Events                         0
dtype: int64

# Challenge 3 - Processing the `Events` Column

#### Our dataframe contains one true text column - the Events column. We should evaluate this column to determine how to process it.

Use the `value_counts()` function to evaluate the contents of this column

In [35]:
austin_fixed['Events'].value_counts()

Events
                             903
Rain                         192
Rain , Thunderstorm          137
Fog , Rain , Thunderstorm     33
Fog                           21
Thunderstorm                  17
Fog , Rain                    14
Rain , Snow                    1
Fog , Thunderstorm             1
Name: count, dtype: int64

Reading the values of `Events` and reflecting what those values mean in the context of data, you realize this column indicates what weather events had happened in a particular day.

#### What is the largest number of events happened in a single day? Enter your answer in the next cell.

In [36]:
austin_fixed.sample(3)

Date  TempHighF  TempAvgF  TempLowF  DewPointHighF  DewPointAvgF  \
767  2016-01-27         59        48        36           30.0          26.0   
463  2015-03-29         84        71        57           60.0          56.0   
344  2014-11-30         79        72        64           63.0          62.0   

     DewPointLowF  HumidityHighPercent  HumidityAvgPercent  \
767          20.0                 70.0                47.0   
463          50.0                 93.0                67.0   
344          59.0                 87.0                71.0   

     HumidityLowPercent  SeaLevelPressureHighInches  \
767                24.0                       30.40   
463                40.0                       30.11   
344                54.0                       30.11   

     SeaLevelPressureAvgInches  SeaLevelPressureLowInches  \
767                      30.32                      30.25   
463                      30.06                      29.99   
344                      30.03                      29.97   

     VisibilityHighMiles  VisibilityAvgMiles  VisibilityLowMiles  WindHighMPH  \
767                 10.0                10.0                10.0         14.0   
463                 10.0                10.0                10.0         17.0   
344                 10.0                10.0                10.0         15.0   

     WindAvgMPH  WindGustMPH Events  
767         5.0         21.0         
463         8.0         28.0         
344         7.0         28.0

In [37]:
austin_fixed['Events'].isna().sum()

0

In [38]:
austin_fixed['Events'].unique()

array(['Rain , Thunderstorm', ' ', 'Rain', 'Fog', 'Rain , Snow',
       'Fog , Rain', 'Thunderstorm', 'Fog , Rain , Thunderstorm',
       'Fog , Thunderstorm'], dtype=object)

In [39]:
espacios_en_blanco = austin_fixed['Events'].eq(" ").sum()
espacios_en_blanco

903

In [40]:
austin_fixed = austin_fixed[austin_fixed['Events'] != " "]


In [41]:
austin_fixed['Events'].unique()

array(['Rain , Thunderstorm', 'Rain', 'Fog', 'Rain , Snow', 'Fog , Rain',
       'Thunderstorm', 'Fog , Rain , Thunderstorm', 'Fog , Thunderstorm'],
      dtype=object)

#### We want to transform the string-type `Events` values to the numbers. This will allow us to apply machine learning algorithms easily.

How? We will create a new column for each type of events (i.e. *Rain*, *Snow*, *Fog*, *Thunderstorm*. In each column, we use `1` to indicate if the corresponding event happened in that day and use `0` otherwise.

Below we provide you a list of all event types. Loop the list and create a dummy column with `0` values for each event in `austin_fixed`. To create a new dummy column with `0` values, simply use `austin_fixed[event] = 0`.

In [42]:
event_list = ['Snow', 'Fog', 'Rain', 'Thunderstorm']

for event in event_list:
    austin_fixed[event] = 0
    austin_fixed.loc[austin_fixed['Events'] == event, event] = 1

austin_fixed

Date  TempHighF  TempAvgF  TempLowF  DewPointHighF  DewPointAvgF  \
0     2013-12-21         74        60        45           67.0          49.0   
18    2014-01-08         53        47        40           51.0          45.0   
19    2014-01-09         70        62        53           60.0          55.0   
20    2014-01-10         72        65        57           64.0          61.0   
22    2014-01-12         67        57        46           58.0          47.0   
...          ...        ...       ...       ...            ...           ...   
1295  2017-07-08        100        88        75           71.0          67.0   
1302  2017-07-15        103        87        71           73.0          70.0   
1304  2017-07-17         98        88        77           75.0          71.0   
1310  2017-07-23        103        90        77           74.0          71.0   
1311  2017-07-24        102        89        76           75.0          71.0   

      DewPointLowF  HumidityHighPercent  HumidityAvgPercent  \
0             43.0                 93.0                75.0   
18            30.0                 93.0                75.0   
19            50.0                 93.0                80.0   
20            54.0                 93.0                81.0   
22            33.0                 84.0                68.0   
...            ...                  ...                 ...   
1295          62.0                 87.0                59.0   
1302          63.0                100.0                65.0   
1304          66.0                 88.0                63.0   
1310          66.0                 85.0                58.0   
1311          56.0                 91.0                60.0   

      HumidityLowPercent  ...  VisibilityAvgMiles  VisibilityLowMiles  \
0                   57.0  ...                 7.0                 2.0   
18                  57.0  ...                 2.0                 1.0   
19                  66.0  ...                 4.0                 0.0   
20                  68.0  ...                 8.0                 1.0   
22                  52.0  ...                10.0                10.0   
...                  ...  ...                 ...                 ...   
1295                31.0  ...                10.0                10.0   
1302                29.0  ...                 9.0                 2.0   
1304                38.0  ...                10.0                10.0   
1310                31.0  ...                10.0                 2.0   
1311                29.0  ...                10.0                 9.0   

      WindHighMPH  WindAvgMPH  WindGustMPH               Events  Snow  Fog  \
0            20.0         4.0         31.0  Rain , Thunderstorm     0    0   
18           12.0         1.0         18.0                 Rain     0    0   
19           12.0         3.0         19.0                  Fog     0    1   
20           13.0         6.0         21.0                 Rain     0    0   
22           15.0         4.0         24.0                 Rain     0    0   
...           ...         ...          ...                  ...   ...  ...   
1295          9.0         3.0         17.0         Thunderstorm     0    0   
1302         23.0         3.0         35.0  Rain , Thunderstorm     0    0   
1304         16.0         3.0         23.0  Rain , Thunderstorm     0    0   
1310         13.0         4.0         24.0  Rain , Thunderstorm     0    0   
1311         13.0         3.0         19.0  Rain , Thunderstorm     0    0   

      Rain Thunderstorm  
0        0            0  
18       1            0  
19       0            0  
20       1            0  
22       1            0  
...    ...          ...  
1295     0            1  
1302     0            0  
1304     0            0  
1310     0            0  
1311     0            0  

[416 rows x 24 columns]

#### Next, populate the actual values in the dummy columns of  `austin_fixed`.

You will check the *Events* column. If its string value contains `Rain`, then the *Rain* column should be `1`. The same for `Snow`, `Fog`, and `Thunderstorm`.

*Hints:*

* Use [`pandas.Series.str.contains()`](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.Series.str.contains.html) to create the value series of each new column.

* What if the values you populated are booleans instead of numbers? You can cast the boolean values to numbers by using `.astype(int)`. For instance, `pd.Series([True, True, False]).astype(int)` will return a new series with values of `[1, 1, 0]`.

In [43]:
for event in event_list:
    austin_fixed[event] = austin_fixed['Events'].str.contains(event).astype(int)


#### Print out `austin_fixed` to check if the event columns are populated with the intended values

In [44]:
austin_fixed.sample(3)

Date  TempHighF  TempAvgF  TempLowF  DewPointHighF  DewPointAvgF  \
207  2014-07-16         95        84        73           74.0          71.0   
314  2014-10-31         76        66        56           61.0          49.0   
102  2014-04-02         82        76        70           69.0          67.0   

     DewPointLowF  HumidityHighPercent  HumidityAvgPercent  \
207          68.0                 93.0                68.0   
314          39.0                 93.0                60.0   
102          63.0                 97.0                78.0   

     HumidityLowPercent  ...  VisibilityAvgMiles  VisibilityLowMiles  \
207                43.0  ...                 8.0                 1.0   
314                27.0  ...                 9.0                 2.0   
102                58.0  ...                 9.0                 4.0   

     WindHighMPH  WindAvgMPH  WindGustMPH               Events  Snow  Fog  \
207         15.0         4.0         22.0  Rain , Thunderstorm     0    0   
314         16.0         7.0         25.0                 Rain     0    0   
102         16.0         9.0         30.0                 Rain     0    0   

     Rain Thunderstorm  
207     1            1  
314     1            0  
102     1            0  

[3 rows x 24 columns]

#### If your code worked correctly, now we can drop the `Events` column as we don't need it any more.

In [45]:
# Your code here


# Challenge 4 - Processing The `Date` Column

The `Date` column is another non-numeric field in our dataset. A value in that field looks like `'2014-01-06'` which consists of the year, month, and day connected with hyphens. One way to convert the date string to numerical is using a similar approach as we used for `Events`, namely splitting the column into numerical `Year`, `Month`, and `Day` columns. In this challenge we'll show you another way which is to use the Python `datetime` library's `toordinal()` function. Depending on what actual machine learning analysis you will conduct, each approach has its pros and cons. Our goal today is to practice data preparation so we'll skip the discussion here.

Here you can find the [reference](https://docs.python.org/3/library/datetime.html) and [example](https://stackoverflow.com/questions/39846918/convert-date-to-ordinal-python) for `toordinal`. The basic process is to first convert the string to a `datetime` object using `datetime.datetime.strptime`, then convert the `datetime` object to numerical using `toordinal`.

#### In the cell below, convert the `Date` column values from string to numeric values using `toordinal()`.

In [46]:
austin_fixed['Date'] = pd.to_datetime(austin_fixed['Date']).apply(lambda x: x.toordinal())


#### Print `austin_fixed` to check your `Date` column.

In [47]:
austin_fixed.head(5)

Date  TempHighF  TempAvgF  TempLowF  DewPointHighF  DewPointAvgF  \
0   735223         74        60        45           67.0          49.0   
18  735241         53        47        40           51.0          45.0   
19  735242         70        62        53           60.0          55.0   
20  735243         72        65        57           64.0          61.0   
22  735245         67        57        46           58.0          47.0   

    DewPointLowF  HumidityHighPercent  HumidityAvgPercent  HumidityLowPercent  \
0           43.0                 93.0                75.0                57.0   
18          30.0                 93.0                75.0                57.0   
19          50.0                 93.0                80.0                66.0   
20          54.0                 93.0                81.0                68.0   
22          33.0                 84.0                68.0                52.0   

    ...  VisibilityAvgMiles  VisibilityLowMiles  WindHighMPH  WindAvgMPH  \
0   ...                 7.0                 2.0         20.0         4.0   
18  ...                 2.0                 1.0         12.0         1.0   
19  ...                 4.0                 0.0         12.0         3.0   
20  ...                 8.0                 1.0         13.0         6.0   
22  ...                10.0                10.0         15.0         4.0   

    WindGustMPH               Events  Snow  Fog  Rain Thunderstorm  
0          31.0  Rain , Thunderstorm     0    0     1            1  
18         18.0                 Rain     0    0     1            0  
19         19.0                  Fog     0    1     0            0  
20         21.0                 Rain     0    0     1            0  
22         24.0                 Rain     0    0     1            0  

[5 rows x 24 columns]

# Challenge 5 - Sampling and Holdout Sets

#### Now that we have processed the data for machine learning, we will separate the data to test and training sets.

We first train the model using only the training set. We check our metrics on the training set. We then apply the model to the test set and check our metrics on the test set as well. If the metrics are significantly more optimal on the training set, then we know we have overfit our model. We will need to revise our model to ensure it will be more applicable to data outside the test set.

#### In the next cells we will separate the data into a training set and a test set using the `train_test_split()` function in scikit-learn.

When using `scikit-learn` for machine learning, we first separate the data to predictor and response variables. This is the standard way of passing datasets into a model in `scikit-learn`. The `scikit-learn` will then find out whether the predictors and responses fit the model.

In the next cell, assign the `TempAvgF` column to `y` and the remaining columns to `X`. Your `X` should be a subset of `austin_fixed` containing the following columns: 

```['Date',
 'TempHighF',
 'TempLowF',
 'DewPointHighF',
 'DewPointAvgF',
 'DewPointLowF',
 'HumidityHighPercent',
 'HumidityAvgPercent',
 'HumidityLowPercent',
 'SeaLevelPressureHighInches',
 'SeaLevelPressureAvgInches',
 'SeaLevelPressureLowInches',
 'VisibilityHighMiles',
 'VisibilityAvgMiles',
 'VisibilityLowMiles',
 'WindHighMPH',
 'WindAvgMPH',
 'WindGustMPH',
 'Snow',
 'Fog',
 'Rain',
 'Thunderstorm']```
 
 Your `y` should be a subset of `austin_fixed` containing one column `TempAvgF`.

In [48]:
austin_fixed.columns

Index(['Date', 'TempHighF', 'TempAvgF', 'TempLowF', 'DewPointHighF',
       'DewPointAvgF', 'DewPointLowF', 'HumidityHighPercent',
       'HumidityAvgPercent', 'HumidityLowPercent',
       'SeaLevelPressureHighInches', 'SeaLevelPressureAvgInches',
       'SeaLevelPressureLowInches', 'VisibilityHighMiles',
       'VisibilityAvgMiles', 'VisibilityLowMiles', 'WindHighMPH', 'WindAvgMPH',
       'WindGustMPH', 'Events', 'Snow', 'Fog', 'Rain', 'Thunderstorm'],
      dtype='object')

In [49]:
X = austin_fixed[['DewPointHighF', 'DewPointAvgF', 'DewPointLowF', 'HumidityHighPercent', 'HumidityAvgPercent', 'HumidityLowPercent', 'SeaLevelPressureHighInches', 'SeaLevelPressureAvgInches' ,'SeaLevelPressureLowInches', 'VisibilityHighMiles', 'VisibilityAvgMiles', 'VisibilityLowMiles', 'WindHighMPH', 'WindAvgMPH', 'Rain', 'Snow', 'Fog', 'Thunderstorm']]
y = austin_fixed['TempAvgF']

In the next cell, import `train_test_split` from `sklearn.model_selection`

In [50]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)


Now that we have split the data to predictor and response variables and imported the `train_test_split()` function, split `X` and `y` into `X_train`, `X_test`, `y_train`, and `y_test`. 80% of the data should be in the training set and 20% in the test set. `train_test_split()` reference can be accessed [here](https://scikit-learn.org/stable/modules/generated/sklearn.model_selection.train_test_split.html).


Enter your code in the cell below:

In [51]:
print(f'100% of our data: {len(austin_fixed)}.')
print(f'80% for training data: {len(X_train)}.')
print(f'20% for test data: {len(X_test)}.')


100% of our data: 416.
80% for training data: 332.
20% for test data: 84.


# Bonus

In [52]:
model = LinearRegression()
model.fit(X_train, y_train)

LinearRegression()

In [53]:
predictions = model.predict(X_test)

In [54]:
print(f'20% for test prediction data is {len(predictions)}.')

20% for test prediction data is 84.


In [55]:
r2 = r2_score(y_test, predictions)
MSE = mean_squared_error(y_test, predictions)
MAE = mean_absolute_error(y_test, predictions)

In [56]:
print("R2 = ", round(r2, 2))
print("The value of the metric MSE is ", round(MSE, 2))
print("MAE = ", round(MAE, 2))

R2 =  0.98
The value of the metric MSE is  2.86
MAE =  1.18


In [57]:
series_pred = pd.Series(predictions, name="predictions_temp")

In [58]:
series_y_test = pd.Series(y_test, name="actual_temp")

In [59]:
comparison_df = pd.concat([series_y_test.reset_index(), series_pred], axis=1)

In [60]:
comparison_df.head(20)

index  actual_temp  predictions_temp
0     483           66         68.724812
1     896           77         77.929520
2    1294           87         86.726129
3     335           68         68.067902
4     265           81         81.188880
5    1253           87         84.831204
6    1156           71         73.489809
7     704           68         70.190145
8     139           76         77.544443
9     194           84         83.524623
10    830           73         73.274720
11   1310           90         86.825883
12    948           84         83.631433
13    104           64         56.190084
14   1262           78         79.491350
15    888           71         71.273451
16    377           41         39.929306
17    707           42         42.791305
18   1008           87         85.663773
19    269           84         84.535600

In [61]:
comparison_df['predictions_temp'] = comparison_df['predictions_temp'].round(decimals=0)


In [62]:
comparison_df.head(20)

index  actual_temp  predictions_temp
0     483           66              69.0
1     896           77              78.0
2    1294           87              87.0
3     335           68              68.0
4     265           81              81.0
5    1253           87              85.0
6    1156           71              73.0
7     704           68              70.0
8     139           76              78.0
9     194           84              84.0
10    830           73              73.0
11   1310           90              87.0
12    948           84              84.0
13    104           64              56.0
14   1262           78              79.0
15    888           71              71.0
16    377           41              40.0
17    707           42              43.0
18   1008           87              86.0
19    269           84              85.0

#### Congratulations! Now you have finished the preparation of the dataset!

# Bonus Challenge 1

#### While the above is the common practice to prepare most datasets, when it comes to time series data, we sometimes do not want to randomly select rows from our dataset.

This is because many time series algorithms rely on observations having equal time distances between them. In such cases, we typically select the majority of rows as the test data and the last few rows as the training data. We don't use `train_test_split()` to select the train/test data because it returns random selections.

In the following cell, compute the number of rows that account for 80% of our data and round it to the next integer. Assign this number to `ts_rows`.

In [63]:
#ts_rows = round(0.8 * len(austin_fixed))

Assign the first `ts_rows` rows of `X` to `X_ts_train` and the remaining rows to `X_ts_test`.

In [64]:
#X_ts_train = X.iloc[:ts_rows, :]
#X_ts_test = X.iloc[ts_rows:, :]


Assign the first `ts_rows` rows of `y` to `y_ts_train` and the remaining rows to `y_ts_test`.

In [65]:
#y_ts_train = y.iloc[:ts_rows]
#y_ts_test = y.iloc[ts_rows:]

In [66]:
#print("Shape of X_ts_train:", X_ts_train.shape)
#print("Shape of X_ts_test:", X_ts_test.shape)
#print("Shape of y_ts_train:", y_ts_train.shape)
#print("Shape of y_ts_test:", y_ts_test.shape)